### PR TITLE
new task to fix stage ordering

### PR DESCRIPTION
**Pull request by ryanai3**

### CODE REVIEW

File: src/main/scala/org/allenai/plugins/DeployPlugin.scala [hunk 1]
```diff
@@ -64,6 +64,8 @@ object DeployPlugin extends AutoPlugin {
       "Cleans the staging directory. This is not done by default by the universal packager."
     )
 
+    val stageAndChecksum = TaskKey[File]("stages and calculates cacheKey of this roject")
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/DeployPlugin.scala]: Missing `p` on `project`.

File: src/main/scala/org/allenai/plugins/DeployPlugin.scala [hunk 2]
```diff
@@ -245,6 +245,12 @@ object DeployPlugin extends AutoPlugin {
     log.info("Deploy complete. Validate your server!")
   }
 
+  val stageAndChecksumTask = stageAndChecksum := {
+    val stageDir = (UniversalPlugin.autoImport.stage in thisProject).value
+    VersionInjectorPlugin.autoImport.injectCacheKey.value
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/DeployPlugin.scala]: This still doesn't fix the problem that the resource is generated in the `resources` folder, but not bundled into the jar.

**Comment by ryanai3** [comment]
@jkinkead pls review

File: src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala [hunk 3]
```diff
@@ -136,49 +126,18 @@ object VersionInjectorPlugin extends AutoPlugin {
...
-        fileName.startsWith("org.allenai.solvers-") ||
-        fileName.startsWith("org.allenai.ari-solvers-")
-    }
-    val hashes = filesToHash
-      .map(Hash.apply)
-      .map(Hash.toHex)
-    // we sort so that we're not dependent on filesystem or git sorting remaining stable in order for the cacheKey
-    // to not change
-    Hash.toHex(Hash((hashes ++ dependentGitCommits.value :+ gitLocalSha1.value).sorted.mkString))
+  lazy val dependentGitCommits: Def.Initialize[Task[Seq[String]]] = Def.taskDyn {
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/VersionInjectorPlugin.scala]: Can you move this to the deploy plugin, and call out to `gitLocalSha1` instead?

File: src/main/scala/org/allenai/plugins/DeployPlugin.scala [hunk 4]
```diff
@@ -245,6 +245,37 @@ object DeployPlugin extends AutoPlugin {
     log.info("Deploy complete. Validate your server!")
   }
 
+  // we stage and generate the cache key in the same task so that stage only runs once
+  val stageAndCacheKeyTask = stageAndCacheKey := {
+    import VersionInjectorPlugin.autoImport.{ gitLocalSha1 }
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/DeployPlugin.scala]: No need for braces if you're only importing one symbol, and not renaming.

File: src/main/scala/org/allenai/plugins/DeployPlugin.scala [hunk 5]
```diff
@@ -245,6 +245,37 @@ object DeployPlugin extends AutoPlugin {
...
+    val stageDir = (UniversalPlugin.autoImport.stage in thisProject).value
+    val logger = streams.value.log
+    val allFiles = ((stageDir / "bin" +++ stageDir / "conf" +++ stageDir / "lib" +++ stageDir / "public") * ".jar").get
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/DeployPlugin.scala]: You're filtering for jars, which we only store in `lib`. I think that's the only directory we want here - other jars won't even end up on our classpath . . .

**Comment by jkinkead** [comment]
A few comments!

**Comment by ryanai3** [comment]
@jkinkead  A few fixes!

File: src/main/scala/org/allenai/plugins/DeployPlugin.scala [hunk 6]
```diff
@@ -245,6 +245,48 @@ object DeployPlugin extends AutoPlugin {
     log.info("Deploy complete. Validate your server!")
   }
 
+  lazy val dependentGitCommits: Def.Initialize[Task[Seq[String]]] = Def.taskDyn {
+    import VersionInjectorPlugin.autoImport.gitLocalSha1
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/DeployPlugin.scala]: Move to the top of the file (especially since it's shared).

**Comment by jkinkead** [comment]
Can you update the deploy plugin README to describe the cache key feature?

**Comment by ryanai3** [comment]
@jkinkead pls review

**Comment by ryanai3** [comment]
@jkinkead ping

**Comment by ryanai3** [comment]
@jkinkead pingerator

File: src/main/scala/org/allenai/plugins/DeployPlugin.scala [hunk 7]
```diff
@@ -245,6 +256,37 @@ object DeployPlugin extends AutoPlugin {
...
+      .map(Hash.toHex)
+    // we sort so that we're not dependent on filesystem or git sorting remaining stable in order for the cacheKey
+    // to not change
+    val cacheKey = Hash.toHex(Hash((hashes ++ dependentGitCommits.value :+ gitLocalSha1.value).sorted.mkString))
+
+    val cacheKeyConfFile = new java.io.File(s"${stageDir.getCanonicalPath}/conf/cacheKey.Sha1")
+
+    logger.info(s"Generating cacheKey.conf managed resource... (cacheKey: $cacheKey)")
+
+    val cacheKeyContents = s"$cacheKey"
```
Comment (jkinkead) [inline src/main/scala/org/allenai/plugins/DeployPlugin.scala]: `cacheKey.toString` - or just `cacheKey` if this is already a string.

**Comment by jkinkead** [comment]
LGTM!

**Pull request merged**